sql/core/src/main/scala/org/apache/spark/sql/execution/CollectMetricsExec.scala

```diff
@@ -69,7 +69,7 @@ case class CollectMetricsExec(
       // - Performance issues due to excessive serialization.
       val updater = collector.copyAndReset()
```

Comment:
Reading the comments above, our goal is to only send back the accumulator updates to the driver side when the task completed, and using a temp task local accumulator doesn't solve the problem when coalesce happens: we call `collector.setState` when finish processing the first partition, but the task is not completed yet.

I think a better solution is to improve the accumulator framework and add a new flag "noUpdateViaHeartbeat", so that the accumulator natively supports this feature and can work with coalesce, and we don't need to create this temp accumulator and can remove the `collector.setState` method.

What do you think? @hvanhovell @Ngone51